### PR TITLE
ssl: Make sure handshake messages are handled correctly during versio…

### LINKED ISSUE #5961: TLS client handshake failure after upgrade to Erlang 24.3.4 on ubuntu
**Describe the bug**

TLS client handshake failure after ansible upgrade of rabbitMQ 3.9.16 and Erlang 24.3.4 on ubuntu.

error in rabbitmq log after attempted login
2022-05-04 16:51:00.992701+02:00 [notice] <0.18735.132> TLS client: In state hello at ssl_handshake.erl:892 generated CLIENT ALERT: Fatal - Handshake Failure
2022-05-04 16:51:00.992701+02:00 [notice] <0.18735.132>  - {unknown_or_malformed_handshake,13}
2022-05-04 16:51:00.998454+02:00 [notice] <0.18738.132> TLS client: In state hello at ssl_handshake.erl:892 generated CLIENT ALERT: Fatal - Handshake Failure
2022-05-04 16:51:00.998454+02:00 [notice] <0.18738.132>  - {unknown_or_malformed_handshake,13}
2022-05-04 16:51:01.004422+02:00 [notice] <0.18741.132> TLS client: In state hello at ssl_handshake.erl:892 generated CLIENT ALERT: Fatal - Handshake Failure
2022-05-04 16:51:01.004422+02:00 [notice] <0.18741.132>  - {unknown_or_malformed_handshake,13}
2022-05-04 16:51:01.010206+02:00 [notice] <0.18746.132> TLS client: In state hello at ssl_handshake.erl:892 generated CLIENT ALERT: Fatal - Handshake Failure
2022-05-04 16:51:01.010206+02:00 [notice] <0.18746.132>  - {unknown_or_malformed_handshake,13}
2022-05-04 16:51:01.015938+02:00 [notice] <0.18742.132> TLS client: In state hello at ssl_handshake.erl:892 generated CLIENT ALERT: Fatal - Handshake Failure
2022-05-04 16:51:01.015938+02:00 [notice] <0.18742.132>  - {unknown_or_malformed_handshake,13}
2022-05-04 16:51:01.021607+02:00 [notice] <0.18736.132> TLS client: In state hello at ssl_handshake.erl:892 generated CLIENT ALERT: Fatal - Handshake Failure
2022-05-04 16:51:01.021607+02:00 [notice] <0.18736.132>  - {unknown_or_malformed_handshake,13}
2022-05-04 16:51:01.021815+02:00 [warning] <0.18732.132> HTTP access denied: rabbit_auth_backend_cache failed authenticating username@hostname: ldap_connect_error

Error persist when upgrading to rabbitmq 3.10.0 

**To Reproduce**
upgrade
RabbitMQ 3.9.10 & Erlang 24.1.7
to
RabbitMQ 3.9.16 & Erlang 24.3.4
with ansible and attempt to login

**Expected behavior**

login directly after upgrade

**Affected versions**

24.3.4

**Additional context**
attaching relevant rabbitmq config
[advanced.config.txt](https://github.com/erlang/otp/files/8621469/advanced.config.txt)
[rabbitmq.conf.txt](https://github.com/erlang/otp/files/8621470/rabbitmq.conf.txt)

### PR DESCRIPTION
**Pull request by IngelaAndin**
…n negotiation

Handshake messages MAY be coalesced into a single TLSPlaintext record
or fragmented across several records. There are also stricter conditions
that must be met for TLS-1.3 "see RFC 8446 section 5.1.  Record Layer"

This means that when supporting both TLS-1.2 and TLS-1.3 as a client we need to
make sure that only the server hello message is handled initially so that
possible coalesced TLS-1.2 handshake messages will be handled by TLS-1.2 code
when the negotiated version has been established.

Closes #5961

### CODE REVIEW

File: lib/ssl/src/tls_gen_connection.erl [hunk 1]
```diff
@@ -389,27 +389,18 @@ handle_protocol_record(#ssl_tls{type = ?APPLICATION_DATA, fragment = Data}, Stat
...
-
-        State = case EffectiveVersion =/= NegotiatedVersion of
-                    State0#state{protocol_buffers =
-                                     Buffers#protocol_buffers{tls_handshake_buffer = Buf}},
+        {Packets, Rest, RecRest} = get_tls_handshakes(Data, StateName, State0),
+            State = State0#state{protocol_buffers = Buffers#protocol_buffers{tls_handshake_buffer = Rest}},
 	case Packets of
             [] -> 
-                assert_buffer_sanity(Buf, Options),
+                #protocol_buffers{tls_handshake_buffer = HSBuffer} = State#state.protocol_buffers,
```
Comment (KennethL) [inline lib/ssl/src/tls_gen_connection.erl:399]: I think the variable `Rest`from line 395 already has the value so no need to dig it out from `State` again.
This line can be omitted if the variable `Rest` is renamed to `HSBuffer` or maybe this line could just be changed to `HSBuffer = Rest`

File: lib/ssl/src/tls_gen_connection.erl [hunk 2]
```diff
@@ -756,21 +765,27 @@ next_record_done(#state{protocol_buffers = Buffers} = State, CipherTexts, Connec
...
 %% When the `negotiated_version` variable is not yet set use the highest supported version.
-effective_version(undefined, #{versions := [Version|_]}, _, _) ->
-    Version;
+handle_unnegotiated_version(undefined, #{versions := [Version|_]} = Options, Data, Buff, _, _) ->
+    tls_handshake:get_tls_handshakes(Version, Data, Buff, Options);
 %% In all other cases use the version saved in the connection state variable `negotiated_version`
-effective_version(Version, _, _, _) ->
-    Version.
+handle_unnegotiated_version(Version, Options, Data, Buff, _, _) ->
+    tls_handshake:get_tls_handshakes(Version, Data, Buff, Options).
```
Comment (KennethL) [inline lib/ssl/src/tls_gen_connection.erl:789]: Seems like handle_unnegotiated_version can return either a 3 tuple or a 2 tuple, please explain why and what is the meaning of returning `{Packet,<<>>, Rest}` at line 779?

Comment (IngelaAndin) [reply lib/ssl/src/tls_gen_connection.erl]: The meaning is that the current handshake buffer is empty and that we have a "Rest" of TLS-record data that we want to parse after handling the first handshake packet.

**Comment by IngelaAndin** [comment]
@KennethL  I made some clarifying changes in the code in my new version of the PR.

**Review by u3s — APPROVED** [review]

**Comment by IngelaAndin** [comment]
I merged my local branch containing this change as I did want to solve the merge conflict via github.

**Pull request closed without merging**